### ISSUE TITLE
[Feature Request] Add Safe Softmax for SDPA

### ISSUE THREAD

**Issue by XXXXRT666** [ml-explore/mlx#2395]
When using SDPA, if a row in the attention mask is fully masked, the output will contain unexpected values — not zeros or NaNs.
This issue arises when performing batched pre-fill for GPT-like models with variable-length inputs with prefix mask and causal mask in a TTS Repo.

<img width="466" height="530" alt="Image" src="https://github.com/user-attachments/assets/053eb599-c45d-45b0-b98e-139f038c61ef" />

```python
import mlx.core as mx
import torch
import torch.nn.functional as F

torch.manual_seed(0)

q = torch.randn(1, 1, 4, 2)
k = torch.randn(1, 1, 4, 2)
v = torch.randn(1, 1, 4, 2)

mask = torch.ones(1, 1, 4, 4, dtype=torch.bool)
mask[0, 0, 1, :] = False

out = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, scale=1)

q_mx = mx.array(q)
k_mx = mx.array(k)
v_mx = mx.array(v)

mask_np = mask.numpy()
mask_mx = mx.array(mask_np)

out_mx = mx.fast.scaled_dot_product_attention(q_mx, k_mx, v_mx, mask=mask_mx, scale=1)


print(mask)
print(mask_mx)
print(out)
print(out_mx)
```

outputs:
```
tensor([[[[ True,  True,  True,  True],
          [False, False, False, False],
          [ True,  True,  True,  True],
          [ True,  True,  True,  True]]]])
array([[[[True, True, True, True],
         [False, False, False, False],
         [True, True, True, True],
         [True, True, True, True]]]], dtype=bool)
tensor([[[[-0.2677, -0.0060],
          [ 0.0000,  0.0000],
          [-0.0891,  0.2425],
          [ 0.0549,  0.4983]]]])
array([[[[-0.267706, -0.00595108],
         [-0.0370273, 0.27384],
         [-0.0890929, 0.242471],
         [0.0548576, 0.498291]]]], dtype=float32)
```

**Comment by XXXXRT666** [comment]
In `torch.device(mps)`,
the output is either 
```
tensor([[[[ True,  True,  True,  True],
          [False, False, False, False],
          [ True,  True,  True,  True],
          [ True,  True,  True,  True]]]], device='mps:0')
array([[[[True, True, True, True],
         [False, False, False, False],
         [True, True, True, True],
         [True, True, True, True]]]], dtype=bool)
tensor([[[[ 0.5999, -0.6573],
          [ 0.0000,  0.0000],
          [ 0.6193, -0.4123],
          [ 0.5115, -0.2631]]]], device='mps:0')
array([[[[0.599913, -0.657345],
         [0.448804, -0.349748],
         [0.619301, -0.412298],
         [0.511456, -0.263086]]]], dtype=float32)
```

or

```
tensor([[[[ True,  True,  True,  True],
          [False, False, False, False],
          [ True,  True,  True,  True],
          [ True,  True,  True,  True]]]], device='mps:0')
array([[[[True, True, True, True],
         [False, False, False, False],
         [True, True, True, True],
         [True, True, True, True]]]], dtype=bool)
tensor([[[[ 0.5999, -0.6573],
          [    nan,     nan],
          [ 0.6193, -0.4123],
          [ 0.5115, -0.2631]]]], device='mps:0')
array([[[[0.599913, -0.657345],
         [0.448804, -0.349748],
         [0.619301, -0.412298],
         [0.511456, -0.263086]]]], dtype=float32)
```
, depending on torch version (2.7.1 stable/ 2.9.0 preview)

**Comment by awni** [comment]
I think `NaN` is the expected result. If you do `softmax([-inf, -inf])` you should get `NaN` (as it's `0/0`). And `NaN` times the value should remain `NaN`.

In fact if you use `-inf` in the mask with MLX you do get `NaN` as expected:

```
import mlx.core as mx

q = mx.random.normal(shape=(1, 1, 1, 2))
k = mx.random.normal(shape=(1, 1, 1, 2))
v = mx.random.normal(shape=(1, 1, 1, 2))

mask = mx.array(-float("inf"))

out = mx.fast.scaled_dot_product_attention(q, k, v, mask=mask, scale=1)
print(out)
```

We should have the same behavior for the boolean mask.

**Comment by XXXXRT666** [comment]
> I think `NaN` is the expected result. If you do `softmax([-inf, -inf])` you should get `NaN` (as it's `0/0`). And `NaN` times the value should remain `NaN`.

Yes, PyTorch is NaNs as well in the past, I think they just replace NaN with 0 in Safe Softmax, and in mlx I think `mx.nan_to_num` could do it for me

> In fact if you use `-inf` in the mask with MLX you do get `NaN` as expected

The problems get solved now, thanks
